Adding version checking to collection

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Make collection check version of Tower compared to the version of the collection and warn on mismatch.

<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - Collection

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 11.0.0
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

john-westcott-iv committed:
Adding version checking to collection

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -434,15 +439,13 @@ def authenticate(self, **kwargs):
         # If we have neither of these, then we can try un-authenticated access
         self.authenticated = True
 
-    def default_check_mode(self):
```
Was this already a defunct method?

john-westcott-iv replied:
Yea, this was from way back in the beginning when I had assumed that every module would be dealing with check mode on its own. The thought was that anything that wanted more than a ping would override this function. But then we figured out how to deal with check mode internally to the class.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -95,6 +97,9 @@ def __init__(self, argument_spec, **kwargs):
         self.session = Request(cookies=CookieJar(), validate_certs=self.verify_ssl)
 
+        # Load the ping page and check the module
+        self.check_version()
```
this is going to add 1 request to every module execution. That's a pretty step performance penalty. Makes me want the API itself to return version data in headers.

john-westcott-iv replied:
If you are going to change anything with this, also add some indicator if its AWX or "Tower".

ryanpetrello replied:
If we want to do this, we should definitely find a way to do it that doesn't involve 2x request overhead.

ryanpetrello replied:
Calculating this:

> add some indicator if its AWX or "Tower".

...per request is a little tricky because determining if it's AWX or not depends on licensing code that can be expensive to run per-request.

ryanpetrello replied:
https://github.com/ansible/awx/pull/6788

john-westcott-iv committed:
Now using new X-API headers to test version and type

beeankha committed:
Update unit tests to not pull in version warnings

Commit:
Remove silence_warning (because of autouse), add version/type warning tests to module_utils unit test file

beeankha commented:
recheck

john-westcott-iv committed:
Attempting to fix py2 test issues

john-westcott-iv committed:
Attempting to fix py2 test issues

beeankha committed:
Add field numbering specification in module_util file, update unit tests

squidboylan committed:
Rework some of our package tooling

beeankha committed:
Fix linter issue

AlanCoding commented on awx_collection/test/awx/test_credential.py:
```diff
@@ -152,7 +152,7 @@ def test_make_use_of_custom_credential_type(run_module, organization, admin_user
 
 
 @pytest.mark.django_db
-def test_secret_field_write_twice(run_module, organization, admin_user, cred_type, silence_warning):
```
this can be removed for _many_ other tests, and I would like to clean that up.

AlanCoding commented on awx_collection/test/awx/test_module_utils.py:
```diff
@@ -1,11 +1,66 @@
+                my_module._COLLECTION_TYPE = "not-junk"
+                my_module.collection_to_version['not-junk'] = 'not-junk'
+                my_module.get_endpoint('ping')
+    mock_warn.assert_called_once_with(
```
here also, the warn method seems to be double-mocked, with `silence_warning` doing the same thing that `mock_warn` does. Not surprised that it works, because the last mock wins, but better to just mock once.

beeankha committed:
Fix up unit tests (no more double mocking)

AlanCoding commented on awx_collection/test/awx/test_module_utils.py:
```diff
@@ -17,18 +70,23 @@ def test_duplicate_config(collection_import):
+
+    class DuplicateTestTowerModule(TowerModule):
+        def load_config(self, config_path):
+            assert config_path == 'my_config'
```
This `load_config` method is still called for other config file locations like `~/.tower_cli.cfg` even if the `tower_config_file` parameter is specified.

Looks like that might not so-much be an issue with the test, but an issue with the expected behavior. Is it expected to _combine_ parameters from the user config file with the specified config file? My current thinking is probably not, but I need others to weigh in. If not, the module utils should have a code change.

AlanCoding commented on awx_collection/tools/roles/template_galaxy/tasks/main.yml:
```diff
@@ -0,0 +1,38 @@
+  replace:
+    path: "{{ collection_path }}/plugins/module_utils/tower_api.py"
+    regexp: '^    _COLLECTION_VERSION =.*'
+    replace: '    _COLLECTION_VERSION = "{{ collection_version }}"'
```
This is going to get a little annoying over time. The development flow is that `make install_collection` is ran before many other important development items can be ran. Yes, there are other replacements made in these tasks, but for `awx.awx`, those are all no-ops. Not this. If you run this you get a diff in your source tree like

```diff
diff --git a/awx_collection/plugins/module_utils/tower_api.py b/awx_collection/plugins/module_utils/tower_api.py
index dc0e14c915..2bd98c1337 100644
--- a/awx_collection/plugins/module_utils/tower_api.py
+++ b/awx_collection/plugins/module_utils/tower_api.py
@@ -33,7 +33,7 @@ class ItemNotDefined(Exception):
 
 class TowerModule(AnsibleModule):
     # This gets set by the make process so whatever is in here is irrelevant
-    _COLLECTION_VERSION = "11.0.0"
+    _COLLECTION_VERSION = "11.2.0"
     _COLLECTION_TYPE = "awx"
     # This maps the collections type (awx/tower) to the values returned by the API
     # Those values can be found in awx/api/generics.py line 204
```

Also, if possible, I would prefer a more static replacement, as opposed to wild cards.

What I'm getting to, I'd prefer to hard code a version like

```python
    _COLLECTION_VERSION = "devel"
```

And use a flag to avoid replacing that, which can be set in `make collection_install`. The actual release pipelines will not care.

I also like that there is _a_ version of the collection which corresponds to the actual source. Yeah, we're doing a lot of build-time replacements, but I would like a sane way of behaving without those.

john-westcott-iv replied:
For the version being coded in, I agree it's a pain because of the source modification. Initially I was hoping that there was a way to get the version of the collection directly from Ansible but I couldn't find anything. Another avenue I explored was trying to have a "version file" somewhere in the collection under a gitignore that the tower_api could load version info out of. Unfortunately, I was having problems locating the file but maybe we can explore that more.

Regarding the wild card replacement I did that somewhat intentionally because of the fact that running `make install_collection` changes  the source. I was attempting to compensate for the case where someone pushes a commit with the value of _COLLECTION_VERSION changed. In the scenario with devel if a PR reviewer misses that line being changed things will get messed up. With the wildcard replacement we don't have that issue.

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -328,6 +338,9 @@ def make_request(self, method, endpoint, *args, **kwargs):
             data = dumps(kwargs.get('data', {}))
 
+        with open('/tmp/john', 'w') as f:
+            f.write("{}".format(self.url.geturl()))
```
I think this still needs removing.

john-westcott-iv replied:
Fixed.

AlanCoding committed:
recover line deletes

AlanCoding commented:
With these changes:

https://github.com/john-westcott-iv/awx/pull/8

I will be good with this.

AlanCoding committed:
Patches from tinkering with tests and default to devel version (#8)

This causes make install_collection to avoid templating the version
  so that it can still be used as-is in development

AlanCoding commented on awx_collection/plugins/module_utils/tower_api.py:
```diff
@@ -374,6 +384,26 @@ def make_request(self, method, endpoint, *args, **kwargs):
+                tower_version = response.getheader('X-API-Product-Version', None)
+            except Exception:
+                tower_type = response.info().getheader('X-API-Product-Name', None)
+                tower_version = response.info().getheader('X-API-Product-Version', None)
```
My only remaining nit-pick is that this will not give a coherent error message when running against a version of AWX or Tower from before the header was introduced.

With the information in variables here, that could be turned into a specifically worded message, but it will involve more moving parts than I would like to put in. So I don't know quite where the balance should be struck, and we also have limited time to spend on the non-recommended use case.
